src/rospkg/common.py

```diff
@@ -45,9 +45,15 @@ class ResourceNotFound(Exception):
     A ROS filesystem resource was not found.
     """
 
-    def __init__(self, msg, ros_paths=None):
+    def __init__(self, msg, ros_paths=None, list_deps_sofar=None):
+        """
+        :type list_deps_sofar: [str]
+        :param list_deps_sofar: List of depended packages at the time the command
+                                               stopped due to this exception.
+        """
         super(ResourceNotFound, self).__init__(msg)
         self.ros_paths = ros_paths
+        self.list_deps_sofar = list_deps_sofar
```

Comment:
The semantic of this field is unclear. It is supposed to contain the unavailable dependencies, right?

Reply:
> It is supposed to contain the unavailable dependencies, right?

No, this argument is for storing the defined dependencies found. I thought this is needed for the usecase where `ResourceNotFound` occurred but the user still wants to get the list of dependencies. If that usecase was not helpful then we could remove this arg (and the related code).

However, I found the current behavior (i.e. w/o this PR) is a bit odd; on the first run of `get_depends`, `ResourceNotFound` occurs, no info about dependencies. This itself might be fine. But then if you run the same command again, all the defined dependency seems to be returned without an exception raised (see the output below). 2nd call succeeds because of the cache, which I think is intended behavior. But it'd be more intuitive if there's a way for the users to get the defined dependency on the 1st run.

```
$ apt-cache policy python-rospkg
python-rospkg:
  Installed: 1.1.4-100
  Candidate: 1.1.4-100
  Version table:
 *** 1.1.4-100 500
        500 http://packages.ros.org/ros/ubuntu xenial/main amd64 Packages
        500 http://packages.ros.org/ros/ubuntu xenial/main i386 Packages
        100 /var/lib/dpkg/status
$ env|grep -i pythonpath
PYTHONPATH=/opt/ros/kinetic/lib/python2.7/dist-packages

$ ipython
In [1]: from rospkg import RosPack
In [2]: rp = RosPack()
In [3]: rp.get_depends("roscpp")
---------------------------------------------------------------------------
ResourceNotFound                          Traceback (most recent call last)
<ipython-input-3-3996ba0273da> in <module>()
----> 1 rp.get_depends("roscpp")

/usr/lib/python2.7/dist-packages/rospkg/rospack.pyc in get_depends(self, name, implicit)
    238
    239             for p in names:
--> 240                 s.update(self.get_depends(p, implicit))
    241             # add in our own deps
    242             s.update(names)

/usr/lib/python2.7/dist-packages/rospkg/rospack.pyc in get_depends(self, name, implicit)
    238
    239             for p in names:
--> 240                 s.update(self.get_depends(p, implicit))
    241             # add in our own deps
    242             s.update(names)

/usr/lib/python2.7/dist-packages/rospkg/rospack.pyc in get_depends(self, name, implicit)
    232
    233             # take the union of all dependencies
--> 234             names = [p.name for p in self.get_manifest(name).depends]
    235
    236             # assign key before recursive call to prevent infinite case

/usr/lib/python2.7/dist-packages/rospkg/rospack.pyc in get_manifest(self, name)
    165             return self._manifests[name]
        166         else:
--> 167             return self._load_manifest(name)
    168
    169     def _update_location_cache(self):

/usr/lib/python2.7/dist-packages/rospkg/rospack.pyc in _load_manifest(self, name)
    209         :raises: :exc:`ResourceNotFound`
    210         """
--> 211         retval = self._manifests[name] = parse_manifest_file(self.get_path(name), self._manifest_name, rospack=self)
    212         return retval
    213

/usr/lib/python2.7/dist-packages/rospkg/rospack.pyc in get_path(self, name)
    201         self._update_location_cache()
    202         if name not in self._location_cache:
--> 203             raise ResourceNotFound(name, ros_paths=self._ros_paths)
    204         else:
    205             return self._location_cache[name]

ResourceNotFound: rosunit
ROS path [0]=/opt/ros/kinetic/share/ros
ROS path [1]=/opt/ros/kinetic/share

In [4]: rp.get_depends("roscpp")
Out[4]:
{'catkin',
 'cpp_common',
 'gencpp',
 'geneus',
 'genlisp',
 'genmsg',
 'gennodejs',
 'genpy',
 'message_generation',
 'message_runtime',
 'roscpp_serialization',
 'roscpp_traits',
 'rostime',
 'std_msgs'}
```

(Btw I also added an argument to store the unavailable dependencies after your comment).